### ISSUE TITLE
Add a new rule to enforce a method syntax

### ISSUE THREAD

**Issue by nstepien** [typescript-eslint/typescript-eslint#1598]
Recently I found out that the shorthand method syntax is looser type-wise than the non-shorthand syntax:

```ts
interface MyInterface {
  fn1(num: number | undefined): void;
  fn2: (num: number | undefined) => void;
}

const badObject: MyInterface = {
  fn1(num: number) {}, // no error??
  fn2(num: number) {} // error
}

const shouldBeBadObject: MyInterface = {
  fn1(num: number) {}, // no error??
  fn2(num: number | undefined) {}
}

const goodObject: MyInterface = {
  fn1(num: number | undefined) {},
  fn2(num: number | undefined) {}
}
```

- [playground](https://www.typescriptlang.org/play/?target=7&ssl=1&ssc=1&pln=19&pc=2#code/JYOwLgpgTgZghgYwgAgLIE8CS5rycgbwChlkYQBGAChAFcBbALmTvoCNpkAfZWkAEwgxQEfgEpmANwD2wfgG4SZEACZmNBs1Yco3XgKEjxyALwA+ZDLmKAvkSIJpIAM5hkbOPwDybAFYQEMGYMbEhYRBQTQiVyalYtBh0xQhsAGmQAegyWaWRoKGkoAH4imNUNJhZE6GSCOzsHJ1dkZwALaVoAG34AIQgezx9-QOCsHHD8KOJSWIqE9hqU9KycvKgC4tKZ8viqhagePkFhEFFa+vtHFzcAc2lpbz8AoLQxsLxI6O24zT2dPSOhlOxjqqTKKjmf04hwMJzOKSINiAA)
- https://github.com/microsoft/TypeScript/issues/36752#issuecomment-585212922

Would it make sense to add a new `@typescript-eslint/eslint-plugin` rule to enforce the non-/shorthand syntax?
